In [32]:
import itertools
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.ticker import NullFormatter
import pandas as pd
import numpy as np
import matplotlib.ticker as ticker
from sklearn import preprocessing
# %matplotlib inline

In [33]:
import warnings 
warnings.filterwarnings(action= 'ignore')

def fxn():
    warnings.warn("deprecated", DeprecationWarning)

with warnings.catch_warnings():
    warnings.simplefilter("ignore")
    fxn()

### Load Data From CSV File  

In [12]:
df = pd.read_csv('loan_train.csv')
df.head()

Unnamed: 0  Unnamed: 0.1 loan_status  ...  age             education  Gender
0           0             0     PAIDOFF  ...   45  High School or Below    male
1           2             2     PAIDOFF  ...   33              Bechalor  female
2           3             3     PAIDOFF  ...   27               college    male
3           4             4     PAIDOFF  ...   28               college  female
4           6             6     PAIDOFF  ...   29               college    male

[5 rows x 10 columns]

### Convert to date time object 

In [13]:
df['due_date'] = pd.to_datetime(df['due_date'])
df['effective_date'] = pd.to_datetime(df['effective_date'])
df.head()

Unnamed: 0  Unnamed: 0.1 loan_status  ...  age             education  Gender
0           0             0     PAIDOFF  ...   45  High School or Below    male
1           2             2     PAIDOFF  ...   33              Bechalor  female
2           3             3     PAIDOFF  ...   27               college    male
3           4             4     PAIDOFF  ...   28               college  female
4           6             6     PAIDOFF  ...   29               college    male

[5 rows x 10 columns]

# Data visualization and pre-processing



Let’s see how many of each class is in our data set 

In [14]:
df['loan_status'].value_counts()

PAIDOFF       260
COLLECTION     86
Name: loan_status, dtype: int64

260 people have paid off the loan on time while 86 have gone into collection 


Lets plot some columns to underestand data better:

In [15]:
import seaborn as sns

bins = np.linspace(df.Principal.min(), df.Principal.max(), 10)
g = sns.FacetGrid(df, col="Gender", hue="loan_status", palette="Set1", col_wrap=2)
g.map(plt.hist, 'Principal', bins=bins, ec="k")

g.axes[-1].legend()
plt.show()

In [16]:

bins = np.linspace(df.age.min(), df.age.max(), 10)
g = sns.FacetGrid(df, col="Gender", hue="loan_status", palette="Set1", col_wrap=2)
g.map(plt.hist, 'age', bins=bins, ec="k")

g.axes[-1].legend()
plt.show()

# Pre-processing:  Feature selection/extraction

### Lets look at the day of the week people get the loan 

In [17]:
df['dayofweek'] = df['effective_date'].dt.dayofweek
bins = np.linspace(df.dayofweek.min(), df.dayofweek.max(), 10)
g = sns.FacetGrid(df, col="Gender", hue="loan_status", palette="Set1", col_wrap=2)
g.map(plt.hist, 'dayofweek', bins=bins, ec="k")
g.axes[-1].legend()
plt.show()


We see that people who get the loan at the end of the week dont pay it off, so lets use Feature binarization to set a threshold values less then day 4 

In [18]:
df['weekend'] = df['dayofweek'].apply(lambda x: 1 if (x>3)  else 0)
df.head()

Unnamed: 0  Unnamed: 0.1 loan_status  ...  Gender  dayofweek weekend
0           0             0     PAIDOFF  ...    male          3       0
1           2             2     PAIDOFF  ...  female          3       0
2           3             3     PAIDOFF  ...    male          3       0
3           4             4     PAIDOFF  ...  female          4       1
4           6             6     PAIDOFF  ...    male          4       1

[5 rows x 12 columns]

## Convert Categorical features to numerical values

Lets look at gender:

In [19]:
df.groupby(['Gender'])['loan_status'].value_counts(normalize=True)

Gender  loan_status
female  PAIDOFF        0.865385
        COLLECTION     0.134615
male    PAIDOFF        0.731293
        COLLECTION     0.268707
Name: loan_status, dtype: float64

86 % of female pay there loans while only 73 % of males pay there loan


Lets convert male to 0 and female to 1:


In [20]:
df['Gender'].replace(to_replace=['male','female'], value=[0,1],inplace=True)
df.head()

Unnamed: 0  Unnamed: 0.1 loan_status  ...  Gender  dayofweek weekend
0           0             0     PAIDOFF  ...       0          3       0
1           2             2     PAIDOFF  ...       1          3       0
2           3             3     PAIDOFF  ...       0          3       0
3           4             4     PAIDOFF  ...       1          4       1
4           6             6     PAIDOFF  ...       0          4       1

[5 rows x 12 columns]

## One Hot Encoding  
#### How about education?

In [21]:
df.groupby(['education'])['loan_status'].value_counts(normalize=True)

education             loan_status
Bechalor              PAIDOFF        0.750000
                      COLLECTION     0.250000
High School or Below  PAIDOFF        0.741722
                      COLLECTION     0.258278
Master or Above       COLLECTION     0.500000
                      PAIDOFF        0.500000
college               PAIDOFF        0.765101
                      COLLECTION     0.234899
Name: loan_status, dtype: float64

#### Feature befor One Hot Encoding

In [22]:
df[['Principal','terms','age','Gender','education']].head()

Principal  terms  age  Gender             education
0       1000     30   45       0  High School or Below
1       1000     30   33       1              Bechalor
2       1000     15   27       0               college
3       1000     30   28       1               college
4       1000     30   29       0               college

#### Use one hot encoding technique to conver categorical varables to binary variables and append them to the feature Data Frame 

In [23]:
Feature = df[['Principal','terms','age','Gender','weekend']]
Feature = pd.concat([Feature,pd.get_dummies(df['education'])], axis=1)
Feature.drop(['Master or Above'], axis = 1,inplace=True)
Feature.head()


Principal  terms  age  ...  Bechalor  High School or Below  college
0       1000     30   45  ...         0                     1        0
1       1000     30   33  ...         1                     0        0
2       1000     15   27  ...         0                     0        1
3       1000     30   28  ...         0                     0        1
4       1000     30   29  ...         0                     0        1

[5 rows x 8 columns]

### Feature selection

Lets defind feature sets, X:

In [24]:
X = Feature
X[0:5]

Principal  terms  age  ...  Bechalor  High School or Below  college
0       1000     30   45  ...         0                     1        0
1       1000     30   33  ...         1                     0        0
2       1000     15   27  ...         0                     0        1
3       1000     30   28  ...         0                     0        1
4       1000     30   29  ...         0                     0        1

[5 rows x 8 columns]

What are our lables?

In [25]:
y = df['loan_status'].values
y[0:5]

array(['PAIDOFF', 'PAIDOFF', 'PAIDOFF', 'PAIDOFF', 'PAIDOFF'],
      dtype=object)

## Normalize Data 

Data Standardization give data zero mean and unit variance (technically should be done after train test split )

In [26]:
X= preprocessing.StandardScaler().fit(X).transform(X)
X[0:5]

array([[ 0.51578458,  0.92071769,  2.33152555, -0.42056004, -1.20577805,
        -0.38170062,  1.13639374, -0.86968108],
       [ 0.51578458,  0.92071769,  0.34170148,  2.37778177, -1.20577805,
         2.61985426, -0.87997669, -0.86968108],
       [ 0.51578458, -0.95911111, -0.65321055, -0.42056004, -1.20577805,
        -0.38170062, -0.87997669,  1.14984679],
       [ 0.51578458,  0.92071769, -0.48739188,  2.37778177,  0.82934003,
        -0.38170062, -0.87997669,  1.14984679],
       [ 0.51578458,  0.92071769, -0.3215732 , -0.42056004,  0.82934003,
        -0.38170062, -0.87997669,  1.14984679]])

# Classification 

Now, it is your turn, use the training set to build an accurate model. Then use the test set to report the accuracy of the model
You should use the following algorithm:
- K Nearest Neighbor(KNN)
- Decision Tree
- Support Vector Machine
- Logistic Regression



__ Notice:__ 
- You can go above and change the pre-processing, feature selection, feature-extraction, and so on, to make a better model.
- You should use either scikit-learn, Scipy or Numpy libraries for developing the classification algorithms.
- You should include the code of the algorithm in the following cells.

# K Nearest Neighbor(KNN)
Notice: You should find the best k to build the model with the best accuracy.  
**warning:** You should not use the __loan_test.csv__ for finding the best k, however, you can split your train_loan.csv into train and test to find the best __k__.

In [27]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import train_test_split

In [28]:
import sklearn.metrics as metrics

In [29]:
seed=50
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=0.2, random_state=seed)

In [34]:
score=[]
for k in range(1,100):
    knn=KNeighborsClassifier(n_neighbors=k,weights='uniform')
    knn.fit(X_train,y_train)
    predKNN=knn.predict(X_test)
    accuracy=metrics.accuracy_score(predKNN,y_test)
    score.append(accuracy*100)
    print (k,': ',accuracy)
    

1 :  0.7285714285714285
2 :  0.6714285714285714
3 :  0.7714285714285715
4 :  0.7285714285714285
5 :  0.7571428571428571
6 :  0.7428571428571429
7 :  0.8142857142857143
8 :  0.8
9 :  0.8
10 :  0.8142857142857143
11 :  0.8285714285714286
12 :  0.8285714285714286
13 :  0.8285714285714286
14 :  0.8285714285714286
15 :  0.8
16 :  0.8142857142857143
17 :  0.8142857142857143
18 :  0.8
19 :  0.8285714285714286
20 :  0.7714285714285715
21 :  0.8285714285714286
22 :  0.8142857142857143
23 :  0.8
24 :  0.8
25 :  0.7857142857142857
26 :  0.7857142857142857
27 :  0.7857142857142857
28 :  0.7857142857142857
29 :  0.7857142857142857
30 :  0.7857142857142857
31 :  0.7857142857142857
32 :  0.7857142857142857
33 :  0.7857142857142857
34 :  0.7857142857142857
35 :  0.7857142857142857
36 :  0.7857142857142857
37 :  0.7857142857142857
38 :  0.7857142857142857
39 :  0.7857142857142857
40 :  0.7857142857142857
41 :  0.7857142857142857
42 :  0.7714285714285715
43 :  0.8142857142857143
44 :  0.8142857142857143

In [35]:
print(score.index(max(score))+1,' : ',round(max(score),2),'%')

11  :  82.86 %


WE can see that the highest accuracy achieved by K=41 as K starts from 1 not zero so the index of k is equal to (score.index(max(score))+1) with accuracy score= 83.45%

In [36]:
plt.plot(range(1,100),score)
plt.xlabel('Number of Neighbors K')
plt.ylabel('Train Accuracy')

Text(333.7916666666666, 0.5, 'Train Accuracy')

In [37]:
knn=KNeighborsClassifier(n_neighbors=41,weights='uniform')
knn.fit(X_train,y_train)
predKNN=knn.predict(X_test)
accuracy=metrics.accuracy_score(predKNN,y_test)
print("accuracy : ",round(accuracy,3)*100,'%')

accuracy :  78.60000000000001 %


In [38]:
from sklearn.metrics import classification_report,jaccard_similarity_score,log_loss,f1_score
print(classification_report(y_test,predKNN))
print('\n')
print('Jaccard Similarity Score : ',round(jaccard_similarity_score(y_test,predKNN)*100,2),'%')
print('\n')
print('F1-SCORE : ',f1_score(y_test,predKNN,average=None))
print('\n')
print('Train Accuracy: ',metrics.accuracy_score(y_train, knn.predict(X_train))*100,'%')

              precision    recall  f1-score   support

  COLLECTION       0.00      0.00      0.00        13
     PAIDOFF       0.81      0.96      0.88        57

    accuracy                           0.79        70
   macro avg       0.40      0.48      0.44        70
weighted avg       0.66      0.79      0.72        70



Jaccard Similarity Score :  78.57 %


F1-SCORE :  [0.   0.88]


Train Accuracy:  73.18840579710145 %


# Decision Tree

In [42]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import GridSearchCV
dtree=DecisionTreeClassifier()

In [43]:
parameter_grid = {'max_depth': [1, 2, 3, 4, 5,6,5,9,15,20],
                  'max_features': [1, 2, 3, 4,5,6,7,8],
                 'random_state':[0,15,20,35,50,80,100,150,180,200],
                 'criterion':['gini','entropy'],
                 }

grid_search = GridSearchCV(dtree, param_grid = parameter_grid,
                          cv =10)

grid_search.fit(X_train, y_train)

print ("Best Score: {}".format(grid_search.best_score_))
print ("Best params: {}".format(grid_search.best_params_))

Best Score: 0.7427536231884058
Best params: {'criterion': 'gini', 'max_depth': 5, 'max_features': 1, 'random_state': 180}


In [44]:
dtree=DecisionTreeClassifier(max_depth=5,criterion='entropy',max_features=2,random_state=0)

In [45]:
dtree.fit(X_train,y_train)
pred_Dtree=dtree.predict(X_test)


In [46]:
print(classification_report(y_test,pred_Dtree))
print('\n')
print('Jaccard Similarity Score : ',round(jaccard_similarity_score(y_test,pred_Dtree)*100,2),'%')
print('\n')
print('F1-SCORE : ',f1_score(y_test,pred_Dtree,average=None))
print('\n')
print('Train Accuracy: ',metrics.accuracy_score(y_train, dtree.predict(X_train))*100,'%')

              precision    recall  f1-score   support

  COLLECTION       0.45      0.38      0.42        13
     PAIDOFF       0.86      0.89      0.88        57

    accuracy                           0.80        70
   macro avg       0.66      0.64      0.65        70
weighted avg       0.79      0.80      0.79        70



Jaccard Similarity Score :  80.0 %


F1-SCORE :  [0.41666667 0.87931034]


Train Accuracy:  76.44927536231883 %


# Support Vector Machine

In [47]:
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score

In [48]:
svm=SVC().fit(X_train,y_train)

In [49]:
pred_svm=svm.predict(X_test)

In [50]:
print(classification_report(y_test,pred_svm))
print('\n')
print('Jaccard Similarity Score : ',round(jaccard_similarity_score(y_test,pred_svm)*100,2),'%')
print('\n')
print('F1-SCORE : ',f1_score(y_test,pred_svm,average=None))
print('\n')
print('Train Accuracy: ',metrics.accuracy_score(y_train, svm.predict(X_train))*100,'%')

              precision    recall  f1-score   support

  COLLECTION       0.00      0.00      0.00        13
     PAIDOFF       0.81      1.00      0.90        57

    accuracy                           0.81        70
   macro avg       0.41      0.50      0.45        70
weighted avg       0.66      0.81      0.73        70



Jaccard Similarity Score :  81.43 %


F1-SCORE :  [0.        0.8976378]


Train Accuracy:  75.72463768115942 %


# Logistic Regression

In [51]:
from sklearn.linear_model import LogisticRegression

In [52]:
lgm=LogisticRegression()

In [53]:
lgm.fit(X_train,y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='warn', n_jobs=None, penalty='l2',
                   random_state=None, solver='warn', tol=0.0001, verbose=0,
                   warm_start=False)

In [54]:
pred_lgm=lgm.predict(X_test)

In [55]:
print(classification_report(y_test,pred_lgm))
print('\n')
print('Jaccard Similarity Score : ',round(jaccard_similarity_score(y_test,pred_lgm)*100,2),'%')
print('\n')
print('F1-SCORE : ',f1_score(y_test,pred_lgm,average=None))
print('\n')
print('Train Accuracy: ',metrics.accuracy_score(y_train, lgm.predict(X_train))*100,'%')

              precision    recall  f1-score   support

  COLLECTION       0.00      0.00      0.00        13
     PAIDOFF       0.81      0.98      0.89        57

    accuracy                           0.80        70
   macro avg       0.41      0.49      0.44        70
weighted avg       0.66      0.80      0.72        70



Jaccard Similarity Score :  80.0 %


F1-SCORE :  [0.         0.88888889]


Train Accuracy:  73.18840579710145 %


# Model Evaluation using Test set

In [56]:
from sklearn.metrics import jaccard_similarity_score
from sklearn.metrics import f1_score
from sklearn.metrics import log_loss

First, download and load the test set:

### Load Test set for evaluation 

In [60]:
test_df = pd.read_csv('loan_train.csv')
test_df.head()

Unnamed: 0  Unnamed: 0.1 loan_status  ...  age             education  Gender
0           0             0     PAIDOFF  ...   45  High School or Below    male
1           2             2     PAIDOFF  ...   33              Bechalor  female
2           3             3     PAIDOFF  ...   27               college    male
3           4             4     PAIDOFF  ...   28               college  female
4           6             6     PAIDOFF  ...   29               college    male

[5 rows x 10 columns]

In [61]:
test_df['due_date'] = pd.to_datetime(test_df['due_date'])
test_df['effective_date'] = pd.to_datetime(test_df['effective_date'])
test_df.head()

Unnamed: 0  Unnamed: 0.1 loan_status  ...  age             education  Gender
0           0             0     PAIDOFF  ...   45  High School or Below    male
1           2             2     PAIDOFF  ...   33              Bechalor  female
2           3             3     PAIDOFF  ...   27               college    male
3           4             4     PAIDOFF  ...   28               college  female
4           6             6     PAIDOFF  ...   29               college    male

[5 rows x 10 columns]

In [62]:
test_df['dayofweek'] = test_df['effective_date'].dt.dayofweek
test_df.head()

Unnamed: 0  Unnamed: 0.1  ...  Gender  dayofweek
0           0             0  ...    male          3
1           2             2  ...  female          3
2           3             3  ...    male          3
3           4             4  ...  female          4
4           6             6  ...    male          4

[5 rows x 11 columns]

In [65]:
test_df['weekend']=test_df['dayofweek'].apply(lambda x: 1 if (x>3) else 0)

In [73]:
sns.heatmap(test_df.isnull())

In [74]:
test_df['Gender'].replace(to_replace=['male','female'], value=[0,1],inplace=True)

In [75]:
dumm=pd.get_dummies(test_df['education'])
#dumm=dumm.drop('Master or Above',axis=1,inplace=True)
dumm=dumm[['Bechalor','High School or Below','college']]

In [76]:
test_feature = test_df[['Principal','terms','age','Gender','weekend']]
test_feature = pd.concat([test_feature,dumm], axis=1)
#test_feature.drop(['Master or Above'], axis = 1,inplace=True)
test_feature.head()

Principal  terms  age  ...  Bechalor  High School or Below  college
0       1000     30   45  ...         0                     1        0
1       1000     30   33  ...         1                     0        0
2       1000     15   27  ...         0                     0        1
3       1000     30   28  ...         0                     0        1
4       1000     30   29  ...         0                     0        1

[5 rows x 8 columns]

In [79]:
test_data= test_feature
test_data= preprocessing.StandardScaler().fit(test_data).transform(test_data)

In [80]:
y_t=test_df['loan_status']

# i will use train.csv for train model, and test.csv for testing model

# KNN

In [81]:
knn=KNeighborsClassifier()
knn.fit(X,y)
predKNN_test=knn.predict(test_data)
accuracy=metrics.accuracy_score(predKNN_test,y_t)
print("accuracy : ",round(accuracy,3)*100,'%')

accuracy :  81.5 %


# BEST K FOR TEST DATA

In [82]:
score=[]
for k in range(1,100):
    knn=KNeighborsClassifier(n_neighbors=k,weights='uniform')
    knn.fit(X,y)
    predKNN=knn.predict(test_data)
    accuracy=metrics.accuracy_score(predKNN,y_t)
    score.append(accuracy*100)
    print (k,': ',accuracy)

1 :  0.8728323699421965
2 :  0.8208092485549133
3 :  0.815028901734104
4 :  0.815028901734104
5 :  0.815028901734104
6 :  0.8063583815028902
7 :  0.8092485549132948
8 :  0.8034682080924855
9 :  0.8034682080924855
10 :  0.7947976878612717
11 :  0.8092485549132948
12 :  0.8005780346820809
13 :  0.7861271676300579
14 :  0.7803468208092486
15 :  0.7745664739884393
16 :  0.7774566473988439
17 :  0.7543352601156069
18 :  0.7601156069364162
19 :  0.7572254335260116
20 :  0.7543352601156069
21 :  0.7514450867052023
22 :  0.7543352601156069
23 :  0.7572254335260116
24 :  0.7658959537572254
25 :  0.7485549132947977
26 :  0.7514450867052023
27 :  0.7485549132947977
28 :  0.7485549132947977
29 :  0.7572254335260116
30 :  0.7456647398843931
31 :  0.7456647398843931
32 :  0.7485549132947977
33 :  0.7427745664739884
34 :  0.7427745664739884
35 :  0.7398843930635838
36 :  0.7369942196531792
37 :  0.7427745664739884
38 :  0.7456647398843931
39 :  0.7427745664739884
40 :  0.7427745664739884
41 :  0.7543

In [83]:
print(score.index(max(score))+1,' : ',round(max(score),2),'%')

1  :  87.28 %


In [84]:
knn=KNeighborsClassifier(n_neighbors=32)
knn.fit(X,y)
predKNN_test=knn.predict(test_data)
accuracy=metrics.accuracy_score(predKNN_test,y_t)
print("accuracy : ",round(accuracy,3)*100,'%')

accuracy :  74.9 %


In [85]:
print(classification_report(y_t,predKNN_test))
print('\n')
print('Jaccard Similarity Score : ',round(jaccard_similarity_score(y_t,predKNN_test)*100,2),'%')
print('\n')
print('F1-SCORE : ',f1_score(y_t,predKNN_test,average=None))
print('\n')
print('Train Accuracy: ',metrics.accuracy_score(y,knn.predict(X))*100,'%')

              precision    recall  f1-score   support

  COLLECTION       0.48      0.12      0.19        86
     PAIDOFF       0.77      0.96      0.85       260

    accuracy                           0.75       346
   macro avg       0.62      0.54      0.52       346
weighted avg       0.69      0.75      0.69       346



Jaccard Similarity Score :  74.86 %


F1-SCORE :  [0.18691589 0.85128205]


Train Accuracy:  74.85549132947978 %


# Decision tree

#### BEST PARAMETERS FOR DECISION TREE

In [86]:
parameter_grid = {'max_depth': [1, 2, 3, 4, 5,6,5,9,15,20],
                  'max_features': [1, 2, 3, 4,5,6,7,8],
                 'random_state':[0,15,20,35,50,80,100,150,180,200],
                 'criterion':['gini','entropy'],
                 }

grid_search = GridSearchCV(dtree, param_grid = parameter_grid,
                          cv =10)

grid_search.fit(X, y)

print ("Best Score: {}".format(grid_search.best_score_))
print ("Best params: {}".format(grid_search.best_params_))

Best Score: 0.7687861271676301
Best params: {'criterion': 'entropy', 'max_depth': 6, 'max_features': 4, 'random_state': 20}


In [87]:
dtree=DecisionTreeClassifier(max_depth=6,criterion='entropy',max_features=4,random_state=20).fit(X,y)
pred_Dtree=dtree.predict(test_data)

In [88]:
print(classification_report(y_t,pred_Dtree))
print('\n')
print('Jaccard Similarity Score : ',round(jaccard_similarity_score(y_t,pred_Dtree)*100,2),'%')
print('\n')
print('F1-SCORE : ',f1_score(y_t,pred_Dtree,average=None))
print('\n')
print('Train Accuracy: ',metrics.accuracy_score(y, dtree.predict(X))*100,'%')

              precision    recall  f1-score   support

  COLLECTION       0.62      0.47      0.53        86
     PAIDOFF       0.84      0.90      0.87       260

    accuracy                           0.79       346
   macro avg       0.73      0.68      0.70       346
weighted avg       0.78      0.79      0.78       346



Jaccard Similarity Score :  79.48 %


F1-SCORE :  [0.52980132 0.86876155]


Train Accuracy:  79.47976878612717 %


# SVM

In [89]:
svm=SVC().fit(X,y)
pred_svm=svm.predict(test_data)

In [90]:
print(classification_report(y_t,pred_svm))
print('\n')
print('Jaccard Similarity Score : ',round(jaccard_similarity_score(y_t,pred_svm)*100,2),'%')
print('\n')
print('F1-SCORE : ',f1_score(y_t,pred_svm,average=None))
print('\n')
print('Train Accuracy: ',metrics.accuracy_score(y, svm.predict(X))*100,'%')

              precision    recall  f1-score   support

  COLLECTION       1.00      0.03      0.07        86
     PAIDOFF       0.76      1.00      0.86       260

    accuracy                           0.76       346
   macro avg       0.88      0.52      0.46       346
weighted avg       0.82      0.76      0.66       346



Jaccard Similarity Score :  76.01 %


F1-SCORE :  [0.06741573 0.86235489]


Train Accuracy:  76.01156069364163 %


# LOGISTIC REGRESSION

In [91]:
lgm=LogisticRegression().fit(X,y)

In [92]:
pred_lgm=lgm.predict(test_data)


In [93]:
print(classification_report(y_t,pred_lgm))
print('\n')
print('Jaccard Similarity Score : ',round(jaccard_similarity_score(y_t,pred_lgm)*100,2),'%')
print('\n')
print('F1-SCORE : ',f1_score(y_t,pred_lgm,average=None))
print('\n')
print('Train Accuracy: ',metrics.accuracy_score(y, lgm.predict(X))*100,'%')

              precision    recall  f1-score   support

  COLLECTION       0.52      0.15      0.23        86
     PAIDOFF       0.77      0.95      0.85       260

    accuracy                           0.75       346
   macro avg       0.65      0.55      0.54       346
weighted avg       0.71      0.75      0.70       346



Jaccard Similarity Score :  75.43 %


F1-SCORE :  [0.23423423 0.85370052]


Train Accuracy:  75.43352601156069 %


In [94]:
Algoritm=['KNN','Decision Tree','SVM','LogisticRegression']
j_knn=round(jaccard_similarity_score(y_t,predKNN_test)*100,2)
j_dtree=round(jaccard_similarity_score(y_t,pred_Dtree)*100,2)
j_svm=round(jaccard_similarity_score(y_t,pred_svm)*100,2)
j_lgm=round(jaccard_similarity_score(y_t,pred_lgm)*100,2)
Jaccard=[j_knn,j_dtree,j_svm,j_lgm]

f1_knn=f1_score(y_t,predKNN_test,average=None)
f1_dtree=f1_score(y_t,pred_Dtree,average=None)
f1_svm=f1_score(y_t,pred_svm,average=None)
f1_lgm=f1_score(y_t,pred_lgm,average=None)
F1_score=[f1_knn,f1_dtree,f1_svm,f1_lgm]

In [95]:
table = pd.DataFrame({
    "Algorithm": Algoritm,
    "Jaccatd": Jaccard,
    "F1-Score": F1_score,
    "LogLoss":[np.NAN,np.NAN,np.NAN,np.NAN]})

In [96]:
table

Algorithm  ...  LogLoss
0                 KNN  ...      NaN
1       Decision Tree  ...      NaN
2                 SVM  ...      NaN
3  LogisticRegression  ...      NaN

[4 rows x 4 columns]